src/analysis/inlay_hints.ml

```diff
@@ -0,0 +1,183 @@
+          { c_rhs = { exp_desc = Texp_let (_, [ {vb_pat; _} ], body); _ }; _ }
+        ]; _}) ->
+        let () = log ~title:"expression" "on function" in
+        let () = iterator.pat iterator vb_pat in
+        iterator.expr iterator body
+      | _ when is_ghost_location avoid_ghost_location expr.exp_loc ->
+        (* Stop iterating when we see a ghost location to avoid
+           annotating generated code *)
+        log ~title:"ghost" "ghost-location found"
+      | _ -> Iterator.default_iterator.expr iterator expr
+  in
+
+  let structure_item_iterator (iterator : Iterator.iterator) item =
+    if overlap_with_loc range item.Typedtree.str_loc then
+      let () = log ~title:"structure_item" "overlap" in
+      match item.str_desc with
+      | Tstr_value (_, bindings) ->
+        List.iter ~f:(fun binding ->
+            expr_iterator iterator binding.Typedtree.vb_expr)
+          bindings
+      | _ when is_ghost_location avoid_ghost_location item.str_loc ->
+        (* Stop iterating when we see a ghost location to avoid
+           annotating generated code *)
+        log ~title:"ghost" "ghost-location found"
+      | _ -> Iterator.default_iterator.structure_item iterator item
+  in
+
+  let pattern_iterator
+      (type a) iterator (pattern : a Typedtree.general_pattern) =
+    let () = log ~title:"pattern" "%a" Logger.fmt (fun fmt ->
+        Format.fprintf fmt "On pattern %a"
+          (Printtyped.pattern 0) pattern
+      )
+    in
+    if overlap_with_loc range pattern.pat_loc
+       && not (pattern_has_constraint pattern)
+    then
+      let () = log ~title:"pattern" "overlap" in
+      let () = Iterator.default_iterator.pat iterator pattern in
+      match pattern.pat_desc with
+      | Tpat_var _ when not pattern.pat_loc.loc_ghost ->
+        let () = log ~title:"pattern" "found" in
+        callback pattern.pat_env pattern.pat_type pattern.pat_loc
+      | _ -> log ~title:"pattern" "not a var"
+  in
+    
+  let iterator = {
+    Ocaml_typing.Tast_iterator.default_iterator with
+    expr = expr_iterator;
+    structure_item = structure_item_iterator;
+    pat = pattern_iterator;
+    value_binding = value_binding_iterator true
+  }
+  in iterator.structure iterator typedtree
+
+type hint = Lexing.position * string
+    
+let create_hint env typ loc =
+  let label =  Printtyp.wrap_printing_env env (fun () ->
+    Format.asprintf "%a" Printtyp.type_scheme typ)
+  in
+  let position = loc.Location.loc_end in
+  (position, label)
+
+let run
+    ~hint_let_binding
+    ~hint_pattern_binding
+    ~avoid_ghost_location
+    ~start
+    ~stop
+    pipeline =
+  let () = log ~title:"start" "%a" Logger.fmt (fun fmt ->
+      Format.fprintf fmt "Start on %s to %s with : let: %b, pat: %b, ghost: %b"
+        (Lexing.print_position () start)
+        (Lexing.print_position () stop)
+        hint_let_binding
+        hint_pattern_binding
+        avoid_ghost_location)
+  in
+  let typer_result = Mpipeline.typer_result pipeline in
```

Comment:
This should move to `Query_command`. The usual separation is that pipeline manipulation is done in query command and tools in analysis are then passed the resulting typedtree (or config, parsetree, etc). That way they are independent of the kernel.

Reply:
Done! You are right, it is better now!